lib/cudnn/upsampling.jl

```diff
@@ -0,0 +1,142 @@
+        rwidth  = output_width  > 1 ? (input_width - 1f0)  / (output_width  - 1f0) : 0f0
+
+        # Compute Y axis lambdas
+        h1r = rheight * out_y
+        h1 = floor(Int, h1r) # here was a typecast (int)
+        h1p = (h1 < input_height - 1) ? 1 : 0
+        h1lambda = h1r - h1
+        h0lambda = 1f0 - h1lambda
+
+        # Compute X axis lambdas
+        w1r = rwidth * out_x
+        w1 = floor(Int, w1r)
+        w1p = (w1 < input_width - 1) ? 1 : 0
+        w1lambda = w1r - w1
+        w0lambda = 1f0 - w1lambda
+
+        for n in 0:num_batch-1  # shift to original C indexing
+            for c in 0:num_channels-1
+                X0 = X[idx(n, num_channels, c, input_height, input_width, h1, w1)]
+                X1 = X[idx(n, num_channels, c, input_height, input_width, h1, w1 + w1p)]
+                X2 = X[idx(n, num_channels, c, input_height, input_width, h1 + h1p, w1)]
+                X3 = X[idx(n, num_channels, c, input_height, input_width, h1 + h1p, w1 + w1p)]
+
+                Y[idx(n, num_channels, c, output_height, output_width, out_y, out_x)] =
+                            h0lambda * (w0lambda * X0 + w1lambda * X1) +
+                            h1lambda * (w0lambda * X2 + w1lambda * X3)
+            end  # channels
+        end  # batch
+    end # 1D kernel loop
+    return nothing
+end
+
+# input is dY, output is dX
+function ∇upsample_bilinear_kernel(
+    input_size,
+    num_channels,
+    input_height,
+    input_width,
+    output_height,
+    output_width,
+    height_scale,
+    width_scale,
+    dY, # const
+    dX)
+    @inbounds for index in ((blockIdx().x - 1) * blockDim().x + threadIdx().x-1): blockDim().x * gridDim().x : input_size-1
+        # mind the order!
+        indexTemp = index
+        in_x = indexTemp % input_width
+        indexTemp ÷= input_width
+        in_y = indexTemp % input_height
+        indexTemp ÷= input_height
+        c = indexTemp % num_channels
+        indexTemp ÷= num_channels
+        n = indexTemp
+
+        out_y = min(in_y / height_scale, output_height - 1)
+        out_x = min(in_x / width_scale, output_width - 1)
+
+        rheight = output_height > 1 ? (output_height - 1.f0) / (input_height - 1.f0) : 0.f0
+        rwidth  = output_width > 1 ?  (output_width -  1.f0) / (input_width -  1.f0) : 0.f0
+
+        # Compute Y axis lambdas
+        h1r = rheight * in_y
+        h1 = round(Int, h1r, RoundDown)
+        h1p = (h1 < output_height - 1) ? 1 : 0
+        h1lambda = h1r - h1
+        h0lambda = 1.f0 - h1lambda
+
+        # Compute X axis lambdas
+        w1r = rwidth * in_x
+        w1 = round(Int, w1r, RoundDown)
+        w1p = (w1 < output_width - 1) ? 1 : 0
+        w1lambda = w1r - w1
+        w0lambda = 1.f0 - w1lambda
+
+        #if __CUDA_ARCH__ >= 350 # true for everything from 9xx on
+        dYi = ldg(dY, index+1)  # ldg(pointer(dY[index])) ?
+        #else
+            # dYi = dY[index + 1];
```

Comment:
`ldg` doesn't have bounds checking, and performance doesn't really matter, so I think its better to switch this around and have the regular access for now. We can optimize later.